### PR TITLE
Add EIP-7623 implementation

### PR DESCRIPTION
**Pull request by nerolation**
Reopened for https://github.com/ethereum/execution-specs/pull/897

### CODE REVIEW

**Comment by SamWilsn** [comment]
This is meant for Prague right? We have our Prague branch at [`forks/prague`](https://github.com/ethereum/execution-specs/tree/forks/prague).

Feel free to move this over there, or I can probably edit the pull request if you'd like.

**Comment by nerolation** [comment]
> This is meant for Prague right? We have our Prague branch at [`forks/prague`](https://github.com/ethereum/execution-specs/tree/forks/prague).
> 
> Feel free to move this over there (both changing the base branch and editing the `src/ethereum/prague` directory instead), or I can probably edit the pull request if you'd like.

How "set" are the changes in src/ethereum/prague?
We haven't had full consensus on the inclusion of 7623 into prague so I'd probably wait until the next ACDE for confirmation.

**Comment by SamWilsn** [comment]
We don't need to merge the pull request until it's ready according to ACDE. I'm suggesting that you make the base branch of your pull request `forks/prague` instead of `master`.

File: src/ethereum/shanghai/fork.py [hunk 1]
```diff
@@ -635,7 +636,13 @@ def process_transaction(
 
     output = process_message_call(message, env)
 
-    gas_used = tx.gas - output.gas_left
+    floor = tokens_in_calldata * TOTAL_COST_FLOOR_PER_TOKEN + TX_BASE_COST
+    if gas_used + tokens_in_calldata * STANDARD_TOKEN_COST  < floor:
```
Comment (SamWilsn) [inline src/ethereum/shanghai/fork.py:640]: Would it be more readable if `gas_used + tokens_in_calldata * STANDARD_TOKEN_COST` was stored in an intermediate variable?

File: src/ethereum/shanghai/fork.py [hunk 2]
```diff
@@ -699,7 +706,7 @@ def validate_transaction(tx: Transaction) -> bool:
     verified : `bool`
         True if the transaction can be executed, or False otherwise.
     """
-    if calculate_intrinsic_cost(tx) > tx.gas:
+    if calculate_intrinsic_cost(tx)[0] > tx.gas:
```
Comment (SamWilsn) [inline src/ethereum/shanghai/fork.py:709]:
```suggestion
    intrinsic_gas = calculate_intrinsic_cost(tx)[0]
    if intrinsic_gas > tx.gas:
```

File: src/ethereum/shanghai/fork.py [hunk 3]
```diff
@@ -731,14 +738,18 @@ def calculate_intrinsic_cost(tx: Transaction) -> Uint:
     -------
     verified : `ethereum.base_types.Uint`
         The intrinsic cost of the transaction.
+    verified : `ethereum.base_types.Uint`
+        The calldata tokens used.
```
Comment (SamWilsn) [inline src/ethereum/shanghai/fork.py:742]: You'll probably want to use better names for the tuple elements.

File: src/ethereum/shanghai/fork.py [hunk 4]
```diff
@@ -751,7 +762,9 @@ def calculate_intrinsic_cost(tx: Transaction) -> Uint:
             access_list_cost += TX_ACCESS_LIST_ADDRESS_COST
             access_list_cost += len(keys) * TX_ACCESS_LIST_STORAGE_KEY_COST
 
-    return Uint(TX_BASE_COST + data_cost + create_cost + access_list_cost)
+    return Uint(TX_BASE_COST +
+                data_cost + 
+                create_cost + access_list_cost), tokens_in_calldata
```
Comment (SamWilsn) [inline src/ethereum/shanghai/fork.py:767]: You'll need to update the return type of the function to `Tuple[Uint, Uint]` I think.

File: src/ethereum/shanghai/transactions.py [hunk 5]
```diff
@@ -20,8 +20,8 @@
 from .fork_types import Address
 
 TX_BASE_COST = 21000
-TX_DATA_COST_PER_NON_ZERO = 16
-TX_DATA_COST_PER_ZERO = 4
+TOTAL_COST_FLOOR_PER_TOKEN = 12
+STANDARD_TOKEN_COST = 4
```
Comment (SamWilsn) [inline src/ethereum/shanghai/transactions.py:24]:
```suggestion
TOTAL_COST_FLOOR_PER_CALLDATA_TOKEN = 12
CALLDATA_TOKEN_COST = 4
```

Unless there's a compelling reason to say "STANDARD" here, I'd prefer not including it.

How do you feel about making it explicit these are calldata related constants?

**Review by SamWilsn — COMMENTED** [review]
Looks like you're still modifying the shanghai fork files, and not prague's

**Comment by nerolation** [comment]
I've merged them here, @SamWilsn :

https://github.com/ethereum/execution-specs/pull/959/files

I'm not sure why the changelog includes other files/changes that I didn't touch.

**Pull request closed without merging**